README.md
S-Compare
=============

NOTE: THIS IS CURRENTLY UNMAINTAINED. EXPECT BUGS.

Python Web Scraping Project - Team S @ University of Essex

S-Compare is a program that collects grocery prices from a number
of major UK retailers, ensures that they are of a standard unit
(per 100 grams for example) and then displays that information in 
a table.

Dependencies:
* Python3
* Tkinter
* Python3-BeautifulSoup4
* Python3-Requests
* Python3-Selenium
* PhantomJS (other browsers can be used but this requires source
modification. For Windows it may also require downloading the
webdriver for the browser in question.)

Platforms known to work:
* GNU/Linux - tested on Arch Linux
* Microsoft Windows - tested on Windows 7

Browser config:
* For GNU/Linux, ensure that you have the browser you wish to use
installed to a standard location such as /usr/bin and ensure that
your distrbution's Selenium package includes the appropriate webdrivers.
* For Microsoft Windows, ensure that the webdriver for the browser
you are using is located either in the program directory or a directory
which is specified in PATH.


CHANGELOG.txt
VERSION: 1.0.0
CHANGES: Downloads html from a given URL, extracts the links and then prints them.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-14

VERSION: 2.0.0
CHANGES: A re-write of version 1.0.0. This version downloads html for the first page of Tesco’s price listing for bottled water. Prices are extracted and then printed.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-26

VERSION: 2.1.0
CHANGES: Improvement on 2.0.0. With this version, the prices are added to a list and the list is then printed.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-26

VERSION: 2.1.5
CHANGES: Improvement on 2.1.0. This version prints a sorted version of the list.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-26

VERSION: 2.9.9
CHANGES: A non-functional re-write of version 2.1.5. The goal of this program was to extract the item titles and well as prices and add them to a separate list.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-26

VERSION: 3.0.0
CHANGES: A re-write of version 2.9.9. This version extracts the prices and item titles and stores them in two lists. The lists are used to create a dictionary, the dictionary is sorted by price then each entry of the dictionary is printed. The following bug is corrected: versions 2.0.0-2.9.9 extracted all prices but not all of these were relevant. This version only extracts the line price.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-11-28

VERSION: 3.0.1
CHANGES: Very minor improvement on 3.0.0. A few lines of code have been reordered so errors involving missing data will be addressed from the very first find operation.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-02

VERSION: 3.9.0
CHANGES: Significant re-write. The 3.0.1 code has been turned into modules: one for Tesco and one for Sainsbury's. More can be added as we go along. The main.py file will import code from the modules and execute it. The code to grab the html has also been outsourced to a module: fetcher.py as it's not possible to have it in main (due to dependency issues) and I don't want to copy the fetch code into the shop modules as that would be rather messy. 
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-05

VERSION: 3.9.1
CHANGES: Bump to 3.9.1 to replace word doc changelog with this text file
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-06

VERSION: 3.9.2
CHANGES: Bump to 3.9.2 to add .txt extension to the license file
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-06

VERSION: 3.9.3
CHANGES: Important fixes for the Tesco and Sainsbury's modules. Scripts for starting the program in Windows and Unix platforms also added.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-10

VERSION: 3.9.4
CHANGES: Code cleanup and improved price sorting.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-11

VERSION: 3.9.5
CHANGES: Renamed starting scripts and made sure that users know from which shop the cheapest item has come from.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-11

VERSION: 3.9.6
CHANGES: Tesco and Sainsbury's modules now fetch prices per 100ml instead of standard line prices. This allows for true price comparison. A couple of minor errors corrected as well.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-12

VERSION: 4.0.0
CHANGES: A full table of prices from all shops, sorted by price, is now output to a file called OUTPUT.txt
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-12

VERSION: 4.1.0
CHANGES: Big improvements to the processing and output code, it has been much simplified. Dictionaries are no longer being as they were not the most suitable way of storing of our data. Also taken code out of functions when it's only being called once. Removed the brackets that were being placed around our prices. The output of lowest prices per shop has also been improved - they're now in neatly formatted tables just the main table.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-13

VERSION: 4.2.0
CHANGES: Overhauled the program's error handling, improved error messages, fixed issue with Sainsbury's title extraction and disabled echo in the windows .bat file for cleaner output.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2014-12-22

VERSION: 4.2.1
CHANGES: Minor bugfix release. Fixed handling of unavailable Tesco products.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-03

VERSION: 4.3.0
CHANGES: Groundwork laid for software to compare prices of all items. Code simplification. Minor fixes.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-04

VERSION: 5.0.0
CHANGES: Software now has the capacity to compare prices for any product accoding to the user's preference. A new function (called dataPull) has been added which allows us to process multiple urls in one go. The urls themselves are stored in a number of text files, categorised by product. The dataPull function will read the urls from the file supplied to it. It will then supply those urls to the relevant shop function.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-04

VERSION: 5.0.1
CHANGES: Minor fix to ensure that the error handling framwork still works with the new program structure.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-04

VERSION: 5.1.0
CHANGES: Table formatting improved. Prices per 75cl instead of 100ml are now convereted (Tesco issue only). Startup message added. Strings of %amp; removed from titles (Sainsbury's only). Important for Sainsburys module with regards to removing unrelated prices.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-08

VERSION: 5.1.1
CHANGES: Fixed another unit issue with Tesco, this time for litres.
CONTRIBUTORS: Charles Bos
RELEASED BY: Charles Bos
DATE: 2015-01-08

VERSION: 5.2.0

main.py
"""
This program fetches prices for common groceries from a number of UK retailers, ensures the prices are of
a standard unit such as per 100 grams, sorts the prices from low to high and then outputs a table of results.

Copyright (C) 2015 Team S

Team S comprises of :
* Charles Bos
* Daniel Bedingfield
* Joshua Coyle
* Oyinpreye Onita
* Sebastian Jakobsen
* Thomas Harris

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
from tesco import tescoData
from sainsburys import sainsburysData
from waitrose import waitroseData
from morrisons import morriData
from tkinter import *
from tkinter import messagebox
from tkinter import filedialog
from threading import Thread
from queue import Queue
import os
import tkinter.ttk as ttk
import webbrowser

class data() :
    def manager(fileName, unit, titleTagEnd, scroll, windowName) :
        '''
        This is the function called by the UI. It coordinates the fetching and displaying
        of results as well as the progressbar. The first 4 args are passed to the call function.
        '''
        # Ensure that only one progressbar window is started for parallel operations.
        # Try except is needed because the object won't be defined on first run.
        try :
            if not UI.runningWinObj.winfo_exists() : UI.runningWin()
        except :
            UI.runningWin()
        
        callThread = Thread(target = data.call, args = (fileName, unit, titleTagEnd, scroll))
        callThread.start() 
        Toplevel.destroy(windowName)

        # This is an ugly hack. Suggestions for better solutions are welcome.
        while True :
            if not callThread.isAlive() :
                data.outputHandler()
                break
            top.update()

    def call(fileName, unit, titleTagEnd, scroll) :
        '''
        This function is run in a separate thread and is called from manager. It calls dataPull to fetch results
        and then processes them. All args are passed to dataPull.
        '''
        tescoResults = data.dataPull('URL_STORE/TESCO/' + fileName, tescoData, 'null', unit, 'null')
        sainsburysResults = data.dataPull('URL_STORE/SAINSBURYS/' + fileName, sainsburysData, '<a href="http://www.sainsburys.co.uk/shop/gb/groceries/' + titleTagEnd, unit, 'null')
        waitroseResults = data.dataPull('URL_STORE/WAITROSE/' + fileName, waitroseData, 'null', unit, scroll)
        morrisonsResults = data.dataPull('URL_STORE/MORRISONS/' + fileName, morriData, 'null', unit, 'null')

        results = [tescoResults, sainsburysResults, waitroseResults, morrisonsResults]
        errors = []
        cheapest = []
        allPrices = []

        for x in results :
            if x[0] != [] : errors += x[0]
            if x[1] != [] :
                cheapest += data.lowestPrices(x[1])
                allPrices += x[1]

        allPricesFormatted = data.createTable(allPrices, "Prices from all shops")
        cheapestFormatted = data.createTable(cheapest, "Lowest prices from each shop")
        priceTable = cheapestFormatted + '\n' + allPricesFormatted

        if errors != [] : utility.writeErrors(errors)
        if allPrices == [] : outputQueue.put(("FullOperationsFailure",))
        elif (allPrices != []) and (errors != []) : outputQueue.put(("PartialOperationsFailure", priceTable))
        else : outputQueue.put((priceTable,))
    
    def dataPull(filePath, shopFunc, titletag, unit, scroll) :
        '''
        This fuction will call the shop module function for each url in a file and then
        return a tuple containing the results.
        '''
        urls = utility.viewFile(filePath)
        urls = urls.split('\n')
        urls = [x for x in urls if x != '']

        errors = []
        prices = []
        
        for x in range(len(urls)) :
            temp = shopFunc(urls[x], titletag, unit, scroll)
            if type(temp) == list : prices += temp
            if type(temp) == str : errors += [temp]

        return (errors, prices)

    def sortPrices(prices) :
        '''
        A bubble sort function for sorting our prices - needed because our price data
        is of type string.
        '''
        x = 0
        y = 1

        for x in range(len(prices)) :
            for y in range(len(prices) -1, x, -1) :
                tupA = prices[x]
                tupB = prices[y]
                priceA = tupA[1]
                priceB = tupB[1]
                endA = priceA.find('/')
                endB = priceB.find('/')
                if float(priceA[1:endA]) > float(priceB[1:endB]) :
                    prices[x] = tupB
                    prices[y] = tupA        

        return prices

    def lowestPrices(prices) :
        prices = data.sortPrices(prices)
        return [list(x) for x in prices if x[1] == prices[0][1]]

    def createTable(prices, tableHeader) :
        prices = data.sortPrices(prices)
        colHeaders = '{:96s}'.format("  Item title") + "Price per unit".ljust(15, ' ') + (' ' * 9) + "Retailer"
        priceTable = str(tableHeader).center(130, ' ') + '\n' + ('=' * len(tableHeader)).center(130, ' ') + '\n' + colHeaders + '\n' + ('-' * 130) + '\n'
        
        for x in range(len(prices)) :
            temp = '{:85s}'.format(prices[x][0]), prices[x][1].rjust(15, ' '), ' ' * 6, prices[x][2]
            priceTable += str(temp)
            priceTable += '\n' + ('-' * 130) + '\n'

        return priceTable.replace('"', ' ').replace("'", ' ').replace(',', ' ').replace('(', ' ').replace(')', ' ')

    def outputHandler() :
        '''
        This function destroys the progressbar window, fetches the processed results from the output
        queue and displays them along with any errors.
        '''
        Toplevel.destroy(UI.runningWinObj)
        output = outputQueue.get() 
        if output[0] == "FullOperationsFailure" : messagebox.showerror(title = "Operation failure", message = "All operations failed. No results to display. Check the logs for errors.")
        elif output[0] == "PartialOperationsFailure" :
            messagebox.showerror(title = "Some errors encountered", message = "Some operations failed. Not all results can be displayed. Check the logs for errors.")
            UI.results(output[1])
        else : UI.results(output[0])

class utility() :
    def contentFetch(funcName, fileName) :
        '''
        An intermediary function for passing data from files to the appropriate part
        of the UI.
        '''
        content = utility.viewFile(fileName)
        if funcName == UI.logViewer :
            if content == 'null' : messagebox.showinfo(title = "Logs", message = "No logs to display.")
            else : UI.logViewer(content)
        else :
            if content == 'null' : messagebox.showerror(title = "Content Failure", message = "Content not found. Please ensure the program has all the necessary files.")
            else : funcName(content)

    def viewFile(fileName) :
        '''
        A general purpose function for reading information from files. Normally it passes its
        contents to the contentFetch function.
        '''
        try :
            file = open(fileName, 'r', encoding = 'utf-8')
            fileText = file.read()
            file.close()
            return fileText
        except IOError :
             return 'null'

    def clearLogs() :
        choice = messagebox.askyesno(title = "Clear Logs?", message = "Would you like to delete the application's logs?")
        if choice == True :
            try :
                os.remove('ERROR_LOG.txt')
            except IOError :
                pass

    def writeErrors(errors) :
        file = open('ERROR_LOG.txt', 'a')
        for x in errors : print(x, file = file)
        file.close()

    def saveFile(content) :
        '''
        Use a Tkinter file dialog to allow the user to save content to a file.
        This is used for the results window in the UI.
        '''
        file = filedialog.asksaveasfile(mode = 'w', defaultextension = '.txt', initialfile = "results")
        if file != None : print(content, file = file)

class UI() :
    # Main window
    def __init__(self, parent) :
        label = Label(parent, text = 'This program compares prices for a number of common groceries. Please select a product category.', wraplength = 345, pady = 5, padx = 5, relief = SUNKEN)
        label.grid(row = 1, column = 1, columnspan = 3)
        button1 = Button(parent, text = "Bread", command = self.bread, height = 5, width = 10).grid(row = 2, column = 1)
        button2 = Button(parent, text = "Dairy", command = self.dairy, height = 5, width = 10).grid(row = 2, column = 2)
        button3 = Button(parent, text = "Crisps and Snacks", command = self.crisps_and_snacks, height = 5, width = 10, wraplength = 80).grid(row = 2, column = 3)
        button4 = Button(parent, text = "Drinks", command = self.drinks, height = 5, width = 10).grid(row = 3, column = 1)
        button5 = Button(parent, text = "Desserts", command = self.desserts, height = 5, width = 10).grid(row = 3, column = 2)
        button6 = Button(parent, text = "Fruit & Veg", command = self.fruit_and_veg, height = 5, width = 10).grid(row = 3, column = 3)
        button7 = Button(parent, text = "Quit", command = parent.destroy).grid(row = 4, column = 3, pady = 10)
        button8 = Button(parent, text = "Homepage", command = lambda : webbrowser.open("https://github.com/charlesbos/s-compare")).grid(row = 4, column = 1, pady = 10)
        button9 = Button(parent, text = "About", command = lambda : utility.contentFetch(self.about, 'ABOUT.txt')).grid(row = 4, column = 2, pady = 10)
        button10 = Button(parent, text = "View Logs", command = lambda : utility.contentFetch(UI.logViewer, 'ERROR_LOG.txt')).grid(row = 5, column = 1, pady = 5)
        button11 = Button(parent, text = "Clear Logs", command = utility.clearLogs).grid(row = 5, column = 2, pady = 5)

    # Product category windows
    def bread(self) :
        bread = Toplevel()
        bread.title("Compare - Bread")
        button1 = Button(bread, text = "Wholemeal Bread", height = 5, width = 12, wraplength = 80, command = lambda : data.manager('BROWN_BREAD.txt', '/100g', 'wholemeal-brown-bread/', 2, bread)).grid(row = 1, column = 1)
        button2 = Button(bread, text = "White Bread", height = 5, width = 12, command = lambda : data.manager('WHITE_BREAD.txt', "/100g", 'white-bread/', 2, bread)).grid(row = 1, column = 2)
        button3 = Button(bread, text = "Baguette", height = 5, width = 12, command = lambda : data.manager('BAGUETTES.txt', '/100g', 'baguettes-part-baked-bread/', 2, bread)).grid(row = 1, column = 3)
        button4 = Button(bread, text = "Rolls", height = 5, width = 12, command = lambda : data.manager('ROLLS.txt', '/each', 'rolls/', 1, bread)).grid(row = 2, column = 1)
        
    def dairy(self) :
        dairy = Toplevel()
        dairy.title("Compare - Dairy")
        button1 = Button(dairy, text = "Milk", height = 5, width = 12, command = lambda : data.manager('MILK.txt', '/100ml', 'fresh-milk/', 5, dairy)).grid(row = 1, column = 1)
        button2 = Button(dairy, text = "Butter", height = 5, width = 12, command = lambda : data.manager('BUTTER.txt', '/100g', 'butter/', 2, dairy)).grid(row = 1, column = 2)
        button3 = Button(dairy, text = "Eggs", height = 5, width = 12, command = lambda : data.manager('EGGS.txt', '/each', 'eggs/', 1, dairy)).grid(row = 1, column = 3)

    def crisps_and_snacks(self) :
        crisps_and_snacks = Toplevel()
        crisps_and_snacks.title("Compare - Crisps & Snacks")
        button1 = Button(crisps_and_snacks, text = "Crisps", height = 5, width = 12, command = lambda : data.manager('CRISPS.txt', '/100g', 'crisps/', 4, crisps_and_snacks)).grid(row = 1, column = 1)
        button2 = Button(crisps_and_snacks, text = "Cereal Bars", height = 5, width = 12, command = lambda : data.manager('CEREAL_BARS.txt', '/100g', 'breakfast-cereal-bars-breakfast-biscuits/', 3, crisps_and_snacks)).grid(row = 1, column = 2)
        button3 = Button(crisps_and_snacks, text = "Popcorn", height = 5, width = 12, command = lambda : data.manager('POPCORN.txt', '/100g', 'popcorn/', 5, crisps_and_snacks)).grid(row = 1, column = 3)

    def drinks(self) :
        drinks = Toplevel()
        drinks.title("Compare - Drinks")
        button1 = Button(drinks, text = "Still Water", height = 5, width = 12, command = lambda : data.manager('STILL_WATER.txt', '/100ml', 'still-water/', 4, drinks)).grid(row = 1, column = 1)
        button2 = Button(drinks, text = "Sparkling Water", height = 5, width = 12, command = lambda : data.manager('SPARKLING_WATER.txt', '/100ml', 'sparkling-water', 2, drinks)).grid(row = 1, column = 2)
        button3 = Button(drinks, text = "Everyday Tea", height = 5, width = 12, command = lambda : data.manager('EVERYDAY_TEA.txt', '/100g', 'everyday-tea/', 3, drinks)).grid(row = 1, column = 3)
        button4 = Button(drinks, text = "Cola", height = 5, width = 12, command = lambda : data.manager('COLA.txt', '/100ml', 'cola/', 4, drinks)).grid(row = 2, column = 1)
        button5 = Button(drinks, text = "Lager", height = 5, width = 12, command = lambda : data.manager('LAGER.txt', '/100ml', 'lager/', 6, drinks)).grid(row = 2, column = 2)
        button6 = Button(drinks, text = "Energy Drink", height = 5, width = 12, command = lambda : data.manager('ENERGY_DRINK.txt', '/100ml', 'energy-drinks/', 3, drinks)).grid(row = 2, column = 3)
        button7 = Button(drinks, text = "Instant Coffee", height = 5, width = 12, command = lambda : data.manager('INSTANT_COFFEE.txt', '/100g', 'instant-coffee/', 5, drinks)).grid(row = 3, column = 1)

    def desserts(self) :
        desserts = Toplevel()
        desserts.title("Compare - Desserts")
        button1 = Button(desserts, text = "Ice Cream Tubs", height = 5, width = 12, command = lambda : data.manager('ICE_CREAM_TUBS.txt', '/100ml', 'ice-cream-tubs/', 6, desserts)).grid(row = 1, column = 1)
        button2 = Button(desserts, text = "Ice Cream Cones", height = 5, width = 12, command = lambda : data.manager('ICE_CREAM_CONES_AND_STICKS.txt', '/100ml', 'lollies-bars-cones/', 4, desserts)).grid(row = 1, column = 2)
        button3 = Button(desserts, text = "Cakes and Tarts", height = 5, width = 12, command = lambda : data.manager('CAKES_AND_TARTS.txt', '/100g', 'frozen-cold-desserts/', 3, desserts)).grid(row = 1, column = 3)

    def fruit_and_veg(self) :
        fruit_and_veg = Toplevel()
        fruit_and_veg.title("Compare - Fruit & Veg")
        button1 = Button(fruit_and_veg, text = "Grapes", height = 5, width = 12, command = lambda : data.manager('GRAPES.txt', '/100g', 'bananas-grapes/', 4, fruit_and_veg)).grid(row = 1, column = 1)
        button2 = Button(fruit_and_veg, text = "Apples", height = 5, width = 12, command = lambda : data.manager('APPLES.txt', '/each', 'apples-pears-rhubarb/', 4, fruit_and_veg)).grid(row = 1, column = 2)
        button3 = Button(fruit_and_veg, text = "Salads", height = 5, width = 12, command = lambda : data.manager('SALADS.txt', '/100g', 'prepared-salad-herbs/', 4, fruit_and_veg)).grid(row = 1, column = 3)
        
    # Other windows
    def results(prices) :
        results = Toplevel()
        results.title("Results")
        frame1 = Frame(results)
        frame2 = Frame(results)
        frame1.pack(fill = BOTH, expand = YES)
        frame2.pack(side = BOTTOM)
        scrollbar = Scrollbar(frame1)
        scrollbar.pack(side = RIGHT, fill = Y)
        text = Text(frame1, height = 40, width = 135, wrap = WORD, yscrollcommand = scrollbar.set)
        scrollbar.config(command = text.yview)
        text.insert(END, prices)
        text.configure(state = DISABLED)
        text.pack(fill = BOTH, expand = YES)
        button1 = Button(frame2, text = "Close", command = results.destroy)
        button1.pack(side = RIGHT)
        button2 = Button(frame2, text = "Save to file", command = lambda : utility.saveFile(prices))
        button2.pack(side = LEFT)

    def about(self, content) :
        about = Toplevel()
        about.title("About")
        frame1 = Frame(about)
        frame2 = Frame(about)
        frame1.pack(fill = BOTH, expand = YES)
        frame2.pack(side = BOTTOM)
        label = Label(frame1, text = content, padx = 5, pady = 5)
        label.pack(fill = BOTH, expand = YES)
        button1 = Button(frame2, text = "License", command = lambda : utility.contentFetch(self.licenseWin, 'LICENSE.txt'))
        button1.pack(side = LEFT)
        button2 = Button(frame2, text = "Changelog", command = lambda : utility.contentFetch(self.changelogWin, 'CHANGELOG.txt'))
        button2.pack(side = LEFT)
        button3 = Button(frame2, text = "Close", command = about.destroy)
        button3.pack(side = RIGHT)

    def licenseWin(self, content) :
        licenseWin = Toplevel()
        licenseWin.title("License")
        frame1 = Frame(licenseWin)
        frame2 = Frame(licenseWin)
        frame1.pack(fill = BOTH, expand = YES)
        frame2.pack(side = BOTTOM)
        scrollbar = Scrollbar(frame1)
        scrollbar.pack(side = RIGHT, fill = Y)
        text = Text(frame1, height = 40, width = 160, wrap = WORD, yscrollcommand = scrollbar.set)
        scrollbar.config(command = text.yview)
        text.insert(END, content)
        text.configure(state = DISABLED)
        text.pack(fill = BOTH, expand = YES)
        button1 = Button(frame2, text = "Close", command = licenseWin.destroy)
        button1.pack(side = TOP)

    def changelogWin(self, content) :
        changelogWin = Toplevel()
        changelogWin.title("Changelog")
        frame1 = Frame(changelogWin)
        frame2 = Frame(changelogWin)
        frame1.pack(fill = BOTH, expand = YES)
        frame2.pack(side = BOTTOM)
        scrollbar = Scrollbar(frame1)
        scrollbar.pack(side = RIGHT, fill = Y)
        text = Text(frame1, height = 40, width = 160, wrap = WORD, yscrollcommand = scrollbar.set)
        scrollbar.config(command = text.yview)
        text.insert(END, content)
        text.configure(state = DISABLED)
        text.pack(fill = BOTH, expand = YES)
        button1 = Button(frame2, text = "Close", command = changelogWin.destroy)
        button1.pack(side = TOP)

    def logViewer(content) :
        logViewer = Toplevel()
        logViewer.title("Logs")
        frame1 = Frame(logViewer)
        frame2 = Frame(logViewer)
        frame1.pack(fill = BOTH, expand = YES)
        frame2.pack(side = BOTTOM)
        scrollbar = Scrollbar(frame1)
        scrollbar.pack(side = RIGHT, fill = Y)
        text = Text(frame1, height = 40, width = 80, wrap = WORD, yscrollcommand = scrollbar.set)
        scrollbar.config(command = text.yview)
        text.insert(END, content)
        text.configure(state = DISABLED)
        text.pack(fill = BOTH, expand = YES)
        button1 = Button(frame2, text = "Close", command = logViewer.destroy)
        button1.pack(side = LEFT)

    def runningWin() :
        UI.runningWinObj = Toplevel()
        UI.runningWinObj.title("Processing")
        UI.runningWinObj.wm_attributes("-topmost", 1)
        frame1 = Frame(UI.runningWinObj)
        frame2 = Frame(UI.runningWinObj)
        frame1.pack()
        frame2.pack(side = BOTTOM)
        label = Label(frame1, text = "Please wait a moment for the results...", pady = 30)
        label.pack()
        progressbar = ttk.Progressbar(frame2, mode = 'indeterminate', length = 350)
        progressbar.pack()
        progressbar.start()

# Initialise queue for data output
outputQueue = Queue()

# Start the interface
top = Tk()
top.title("S-Compare")
ui = UI(top)
top.mainloop()


extra.py
"""
extra.py

A module which contains functions for extra functionality such as supplying text to dialogues or creating error logs.
"""
import os
from tkinter import messagebox

def viewFile(fileName) :
    '''
    A function to read content from files and return either that content
    or else a keyword to indicate that the procedure failed.
    One argument taken, the name of the file form which to read the content.
    '''
    try :
        file = open(fileName, 'r')
        fileText = file.read()
        file.close()
        return fileText
    except IOError :
         return 'null'

def errorLog(time, message, listLengths, exceptDetails) :
    '''
    A function to write errors to a text file. Depending on the nature of the
    error, this function will write the following: time and date, lengths of
    price and item lists (if they are unequal), a programmer defined message
    and the details of the Python exception that was raised (if one was raised).
    NOTE: more than one function may call the this function, each supplying it's own
    bit of information.
    Four arguments are taken: the time and date (as formatted in strftime), the programmer
    defined message, the list lengths and the exception details. Arguments of 'null' are
    ignored.
    '''
    file = open('ERROR_LOG.txt', 'a')
    if time != 'null' : print(time, file = file)
    if exceptDetails != 'null' : print(exceptDetails, file = file)
    if listLengths != 'null' : print(listLengths, file = file)
    if message != 'null' : print(message + '\n' + ('-' * 100), file = file)
    file.close()

def clearLogs() :
    '''
    A function which uses the tkinter messagebox to ask the user if they
    would like to clear the logs. If the answer is yes, the program will
    attempt to delete the error log file.
    No arguments taken.
    '''
    choice = messagebox.askyesno(title = "Clear Logs?", message = "Would you like to delete the application's logs?")
    if choice == True :
        try :
            os.remove('ERROR_LOG.txt')
        except :
            pass
        
    

    
    


morrisons.py
"""
This program extracts prices and item titles from a valid Morrisons store page
url and then returns that data.

Copyright (C) 2015 Team S

Team S comprises of :
* Charles Bos
* Daniel Bedingfield
* Joshua Coyle
* Oyinpreye Onita
* Sebastian Jakobsen
* Thomas Harris

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
from fetcher import simpleFetch
from time import strftime

def morriData(url, titletag, unit, scroll) :
    '''
    Extracts a list of item prices and a list of item titles. Will try and return a
    tuple containing the aforementioned lists plus a list containing the shop name.
    Titletag and scroll args are not used and are specified for compatibilty reasons only.
    '''
    htmlString = simpleFetch(url)
    
    if htmlString == "null" :
        errorTime = strftime('%H:%M:%S %Y-%m-%d')
        errorMessage = "MorrisonsError: failed to retrieve webpage."
        return errorTime + '\n' + errorMessage + '\n' + url + '\n' + '-' * 80 
    else :
        # Extract prices
        priceList = []
        priceStart = htmlString.find('<p class="pricePerWeight">')
        priceEnd = htmlString.find('</p>', priceStart)

        while 0 <= priceStart <= len(htmlString) :
            priceExtract = htmlString[priceStart + 26:priceEnd]
            if len(priceExtract) > 0 :
                priceExist = htmlString.find('Out of stock', priceEnd, priceEnd + 500)
                if priceExtract.find('p ') != -1 : priceMeasureStart = priceExtract.find('p ')
                elif priceExtract.find(' p') != -1 : priceMeasureStart = priceExtract.find(' p')
                priceMeasure = priceExtract[priceMeasureStart:priceEnd]
                if priceExtract[0] == '£' : priceExtract = priceExtract[1:]
                if priceMeasure.find('p ') != -1 :
                    priceExtract = '{:.2f}'.format(float(priceExtract[:priceMeasureStart]) / 100)
                if (priceMeasure.find("litre") != -1) or (priceMeasure.find("kg") != -1) :
                    priceExtract = '{:.2f}'.format(float(priceExtract[:priceMeasureStart]) / 10)
                if priceMeasure.find("75cl") != -1 :
                    priceExtract = '{:.2f}'.format((float(priceExtract[:priceMeasureStart]) / 30) * 4)
                priceExtract = '£' + priceExtract + unit
                if priceExist == -1 : priceList += [priceExtract]
            priceStart = htmlString.find('<p class="pricePerWeight">', priceEnd)
            priceEnd = htmlString.find('</p>', priceStart)

        # Extract titles
        titleList = []
        titleStart = htmlString.find('<strong itemprop="name">')
        titleEnd = htmlString.find('</strong>', titleStart)

        while 0 <= titleStart <= len(htmlString) :
            if htmlString.find('<p class="pricePerWeight"> </p>', titleEnd, titleEnd + 3000) == -1 :
                measureEnd = htmlString.find('</a>', titleEnd)
                measure = htmlString[titleEnd:measureEnd].partition('\n')[-1].partition('             ')[-1].partition('\r\n')[0]
                titleExtract = htmlString[titleStart + 24:titleEnd].replace('amp;', '').lstrip('<abbr title= ').partition('>')[0].strip('''"''')
                titleExtract = titleExtract + ' ' + measure
                titleExist = htmlString.find('Out of stock', titleEnd, titleEnd + 1500)
                if titleExist == -1 : titleList += [titleExtract]
            titleStart = htmlString.find('<strong itemprop="name">', titleEnd)
            titleEnd = htmlString.find('</strong>', titleStart)

        if len(priceList) != len(titleList) :
                errorTime = strftime('%H:%M:%S %Y-%m-%d')
                errorMessage = "MorrisonsError: lengths of prices and item titles do not match."
                listLengths = 'priceList length = ' + str(len(priceList)) + '\n' + 'titleList length = ' + str(len(titleList))
                return errorTime + '\n' + errorMessage + '\n' + listLengths + '\n' + url + '\n' + '-' * 80
        else : return [list(x) for x in zip(titleList, priceList, ["Morrisons"] * len(priceList))]
